idx/memory/tag_query.go

```diff
@@ -0,0 +1,408 @@
+
+		var shortCut bool
+		var re *regexp.Regexp
+		var err error
+
+		// shortcut if pattern is ^.+, any value will match so there's no need
+		// to actually run the regular expression.
+		// (f.e. expression "key!=" will be translated to "key=~^.+")
+		if e.value == "^.+" {
+			shortCut = true
+		} else {
+			re, err = regexp.Compile(e.value)
+			if err != nil {
+				return errInvalidQuery
+			}
+		}
+
+		// cache all tags that have once matched the regular expression.
+		// this is based on the assumption that many matching tags will be repeated
+		// over multiple series, so there's no need to run the regex for each of them
+		// because once we know that a tag matches we can just compare strings
+		matchingTags := make(map[string]struct{})
+	IDS:
+		for id := range resultSet {
+			var def *idx.Archive
+			var ok bool
+			if def, ok = byId[id.ToString()]; !ok {
+				// should never happen because every ID in the tag index
+				// must be present in the byId lookup table
+				CorruptIndex.Inc()
+				delete(resultSet, id)
+				continue IDS
+			}
+
+			for _, tag := range def.Tags {
+				// reduce regex matching by looking up cached matches
+				if _, ok := matchingTags[tag]; ok {
+					if not {
+						delete(resultSet, id)
+					}
+					continue IDS
+				}
+			}
+
+			for _, tag := range def.Tags {
+				tagSplits := strings.SplitN(tag, "=", 2)
+				if len(tagSplits) != 2 {
+					// should never happen because every tag in the index
+					// must have a valid format
+					InvalidTagInIndex.Inc()
+					delete(resultSet, id)
+					continue IDS
+				}
+
+				if e.key == tagSplits[0] && (shortCut || re.MatchString(tagSplits[1])) {
+					matchingTags[tag] = struct{}{}
+					if not {
+						delete(resultSet, id)
+					}
+					continue IDS
+				}
+			}
+			if !not {
+				delete(resultSet, id)
+			}
+		}
+	}
+	return nil
+}
+
+func (q *TagQuery) filterByFrom(resultSet TagIDs, byId map[string]*idx.Archive) {
+	if q.from <= 0 {
+		return
+	}
+
+	for id := range resultSet {
+		var def *idx.Archive
+		var ok bool
+		if def, ok = byId[id.ToString()]; !ok {
+			// should never happen because every ID in the tag index
+			// must be present in the byId lookup table
+			CorruptIndex.Inc()
+			delete(resultSet, id)
+			continue
+		}
+
+		if def.LastUpdate < q.from {
+			delete(resultSet, id)
+		}
+	}
+}
+
+func (q *TagQuery) Run(index TagIndex, byId map[string]*idx.Archive) (TagIDs, error) {
+	var skipMatch, skipEqual = -1, -1
+	var resultSet TagIDs
+	var err error
+
+	// find the best expression to start with and retrieve its resultSet
+	if len(q.equal) == 0 {
+		if len(q.match) == 0 {
+			return nil, errInvalidQuery
+		}
+
+		skipMatch, resultSet, err = q.getInitialByMatch(index)
+		if err != nil {
+			return nil, errInvalidQuery
+		}
+	} else {
+		skipEqual, resultSet = q.getInitialByEqual(index)
+	}
+
+	// filter the resultSet by the from condition and all other expressions given.
+	// the order of those filters should to be increasing by the cpu required
```

Comment:
Sentence is a little funky. Maybe, "Filters should be in ascending order by the cpu required"